### PR TITLE
Cleanup IotHubConnectionStringBuilder tests

### PR DESCRIPTION
**Pull request by drwill-ms**

### CODE REVIEW

File: iothub/device/src/AuthenticationWithTokenRefresh.cs [hunk 1]
```diff
@@ -22,21 +22,6 @@ public abstract class AuthenticationWithTokenRefresh : IAuthenticationMethod, ID
         private string _token;
         private bool _isDisposed;
 
-        /// <summary>
```
Comment (drwill-ms) [inline iothub/device/src/AuthenticationWithTokenRefresh.cs:25]: ctor above properties

File: e2e/test/iothub/DeviceTokenRefreshE2ETests.cs [hunk 2]
```diff
@@ -245,7 +245,7 @@ public Task WaitForTokenRefreshAsync(CancellationToken cancellationToken)
                 return _tokenRefreshSemaphore.WaitAsync(cancellationToken);
             }
 
-            protected override Task<string> SafeCreateNewToken(string iotHub, int suggestedTimeToLive)
+            protected override Task<string> SafeCreateNewTokenAsync(string iotHub, int suggestedTimeToLive)
```
Comment (drwill-ms) [inline e2e/test/iothub/DeviceTokenRefreshE2ETests.cs:248]: Async suffix for method that turn a Task

File: iothub/device/src/DeviceAuthenticationWithRegistrySymmetricKey.cs [hunk 3]
```diff
@@ -37,8 +38,8 @@ public string DeviceId
         /// <summary>
         /// Gets or sets the key associated with the device.
         /// </summary>
-        /// 
-        [System.Diagnostics.CodeAnalysis.SuppressMessage(
+        ///
```
Comment (abhipsaMisra) [inline iothub/device/src/DeviceAuthenticationWithRegistrySymmetricKey.cs:41]: nit: extra comment

File: iothub/device/src/DeviceAuthenticationWithSakRefresh.cs [hunk 4]
```diff
@@ -11,18 +11,18 @@ namespace Microsoft.Azure.Devices.Client
...
-        private IotHubConnectionString _connectionString;
+        private readonly IotHubConnectionString _connectionString;
 
         public DeviceAuthenticationWithSakRefresh(
-            string deviceId, 
+            string deviceId,
             IotHubConnectionString connectionString) : base(deviceId)
         {
-            _connectionString = connectionString;
+            _connectionString = connectionString ?? throw new ArgumentNullException(nameof(connectionString));
```
Comment (abhipsaMisra) [inline iothub/device/src/DeviceAuthenticationWithSakRefresh.cs:20]: why do we not want to throw this exception?

Comment (drwill-ms) [reply iothub/device/src/DeviceAuthenticationWithSakRefresh.cs:20]: I added the throw. What do you mean?

File: iothub/device/src/HsmAuthentication/ModuleAuthenticationWithHsm.cs [hunk 5]
```diff
@@ -21,25 +21,27 @@ internal class ModuleAuthenticationWithHsm : ModuleAuthenticationWithTokenRefres
         /// <param name="signatureProvider">Provider for the token signature.</param>
         /// <param name="deviceId">Device Identifier.</param>
         /// <param name="moduleId">Module Identifier.</param>
-        public ModuleAuthenticationWithHsm(ISignatureProvider signatureProvider, string deviceId, string moduleId, string generationId) : base(deviceId, moduleId)
+        /// <param name="generationId"></param>
```
Comment (drwill-ms) [inline iothub/device/src/HsmAuthentication/ModuleAuthenticationWithHsm.cs:24]: What should this comment be?

@abhipsaMisra ?

File: iothub/device/src/DeviceAuthenticationWithTokenRefresh.cs [hunk 6]
```diff
@@ -51,6 +46,11 @@ public DeviceAuthenticationWithTokenRefresh(string deviceId)
             DeviceId = deviceId;
         }
 
+        /// <summary>
+        /// Gets the DeviceID.
```
Comment (azabbasi) [inline iothub/device/src/DeviceAuthenticationWithTokenRefresh.cs:50]: Device Id

File: iothub/device/tests/ConnectionString/IotHubConnectionStringBuilderTests.cs [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+        private const string DeviceIdSplChar = "device1-.+%_#*?!(),=@;$'";
+        private const string ModuleId = "moduleId";
+        private const string SharedAccessKey = "dGVzdFN0cmluZzE=";
+        private const string SharedAccessKeyName = "AllAccessKey";
+        private const string CredentialScope = "Device";
+        private const string CredentialType = "SharedAccessSignature";
+        private const string SharedAccessSignature = "SharedAccessSignature sr=dh%3a%2f%2facme.azure-devices.net&sig=dGVzdFN0cmluZzU=&se=87824124985&skn=AllAccessKey";
+
+        [TestMethod]
+        public void IotHubConnectionStringBuilder_ParamConnectionString_ParsesHostName()
```
Comment (drwill-ms) [inline iothub/device/tests/ConnectionString/IotHubConnectionStringBuilderTests.cs:29]: I broke the one long test out into separate cases, so we more easily track what is being tested, and what the gaps may be.

File: iothub/device/src/HsmAuthentication/ModuleAuthenticationWithHsm.cs (hunk 5 above)
Comment (abhipsaMisra) [reply iothub/device/src/HsmAuthentication/ModuleAuthenticationWithHsm.cs:24]: "Generation Id for the Edge module"?
Not that I know what it represents, we use it here: https://github.com/Azure/azure-iot-sdk-csharp/blob/master/iothub/device/src/Edge/EdgeModuleClientFactory.cs#L27

File: iothub/device/src/DeviceAuthenticationWithSakRefresh.cs (hunk 4 above)
Comment (azabbasi) [reply iothub/device/src/DeviceAuthenticationWithSakRefresh.cs:20]: I don't understand the question @abhipsaMisra :D

File: iothub/device/src/HsmAuthentication/ModuleAuthenticationWithHsm.cs (hunk 5 above)
Comment (abhipsaMisra) [reply iothub/device/src/HsmAuthentication/ModuleAuthenticationWithHsm.cs:24]: https://github.com/Azure/azure-iot-sdk-csharp/blob/master/iothub/device/src/Edge/EdgeModuleClientFactory.cs#L96

Comment (drwill-ms) [reply iothub/device/src/HsmAuthentication/ModuleAuthenticationWithHsm.cs:24]: We don't know what generation Id is?

**Review by azabbasi — APPROVED** [review]

File: iothub/device/src/HsmAuthentication/ModuleAuthenticationWithHsm.cs (hunk 5 above)
Comment (drwill-ms) [reply iothub/device/src/HsmAuthentication/ModuleAuthenticationWithHsm.cs]: I found it looking at the IoT Hub data plane swagger.

**Review by abhipsaMisra — APPROVED** [review]

File: iothub/device/src/IotHubConnectionStringBuilder.cs [hunk 8]
```diff
@@ -165,8 +166,17 @@ public IAuthenticationMethod AuthenticationMethod
         /// <summary>
         /// Gets the shared access signature used to connect to the IoT Hub service.
         /// </summary>
+        /// <remarks>
+        /// This is used when a device app creates its own limited-lifespan SAS token, instead of letting
+        /// this SDK derive one from a shared access token. When a device client is initialized with a
+        /// SAS token, when that token expires, the client must be disposed, and if desired, recreated
+        /// with a newly dervied SAS token.
```
Comment (barustum) [inline iothub/device/src/IotHubConnectionStringBuilder.cs:173]: derived

**Comment by barustum** [comment]
apologies for the late review. I made one comment

**Review by barustum — APPROVED** [review]

**Pull request merged**